src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -30,3 +60,171 @@ prepLookup :: Run fd -> SerialisedKey -> Maybe PageSpan
 prepLookup (_fd, b, fpix) k
   | Bloom.elem k b = Index.toPageSpan $ Index.search k fpix
   | otherwise      = Nothing
+
+{-------------------------------------------------------------------------------
+  Lookups in IO
+-------------------------------------------------------------------------------}
+
+{-
+  Note [Batched lookups, buffer strategy and restrictions]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  As a first implementation, we use a simple buffering strategy for batched
+  lookups: allocate fresh buffers, and GC them once they are no longer used. In
+  a later phase of development, we will look into more elaborate buffer
+  strategies, for example using a cache of page buffers, which reduces the
+  number of blocks of memory that we allocate/free.
+
+  When we implement a reusable buffer strategy, we will have to take extra care
+  to copy bytes from raw pages when necessary. 'rawPageLookup' slices serialised
+  values from pages without copying bytes. If the serialised value survives
+  after the reusable buffer is returned to the cache, accessing that serialised
+  value's bytes has undefined behaviour __unless__ we ensure that we copy from
+  the raw page instead of or in addition to slicing.
+
+  There are currently no alignment constraints on the buffers, but there will be
+  in the future. The plan is to optimise file access by having key\/operation
+  files opened with the @O_DIRECT@ flag, which requires buffers to be aligned to
+  the filesystem block size (typically 4096 bytes). We expect to benefit from
+  this flag in the UTxO use case, because UTxO keys are uniformly distributed
+  hashes, which means that there is little to no spatial locality when
+  performing lookups, and so we can skip looking at the page cache.
+-}
+
+newtype BatchSize = BatchSize { unBatchSize :: Int }
+
+-- | A batch lookup read fewer bytes than requested.
+data NotEnoughBytes = NotEnoughBytes
+  deriving (Show, Exception)
+
+{-# SPECIALIZE lookupsInBatches ::
+       HasBlockIO IO HandleIO
+    -> Resolve1
+    -> BatchSize
+    -> [Run (Handle HandleIO)]
+    -> [SerialisedKey]
+    -> IO [(SerialisedKey, Maybe (Entry SerialisedValue BlobSpan))]
+  #-}
+-- | Batched lookups.
+--
+-- When the length of the list of lookup keys exceeds the batch size, the list
+-- is subdivided into batches of at most the given batch size. Each batch is
+-- submitted concurrently using async IO, which could result in a speedup if
+-- this code is run with multi-threading enabled.
+--
+-- See Note [Batched lookups, buffer strategy and restrictions]
+--
+-- TODO: optimise by reducing allocations, possibly looking at core, or
+-- revisiting the batching approach.
+lookupsInBatches ::
+     forall m h. (MonadAsync m, PrimMonad m, MonadThrow m)
+  => HasBlockIO m h
+  -> Resolve1
+  -> BatchSize
+  -> [Run (Handle h)]
+  -> [SerialisedKey]
+  -> m [(SerialisedKey, Maybe (Entry SerialisedValue BlobSpan))]
```

Comment:
Returning only the `BlobSpan` here is problematic, since reading the actual blob requires knowing which blob file (or run) to read from, but that information is lost at this point. So we might want to return a `BlobRef`, or some internal reference type that can be turned into a `BlobRef` later.

Reply:
I don't know what the external/public `BlobRef` type should look like, but for the internal one to be used here (and in run merging) I don't see the need to index it by `m` or its "type" `blob`. Here we are dealing with serialised plain bytes. Everything else comes in the layer above.

Maybe something like `data BlobRef fhandle = BlobRef !(Run.Run fhandle) !BlobSpan`, defined locally in this module? That's what I use for run merging, too, but we can still pull it into a shared definition once we're sure the same representation really makes sense in both places.

Reply:
I agree, the `m` type is probably not necessary, and we should instead index by the file handle type

It would be an advantage if we can unify the external and internal `BlobRef` types, so we don't have to do additional traversals of lookup results to convert internal `BlobRef` types to external `BlobRef` types. It sounds like `data BlobRef fhandle !(Run.Run fhandle) !blob` would work for your case too

Reply:
So the `fhandle` would be the same one that is inside the run? Feels a bit redundant.

Reply:
That was a typo :P it doesn't need that one

Reply:
> It would be an advantage if we can unify the external and internal BlobRef types, so we don't have to do additional traversals of lookup results to convert internal BlobRef types to external BlobRef types.

So what is going to get returned to the user? I think I remember the users themselves have to use the reference to fetch the `SerialisedBlob` (they might only want to pay that cost for a few select ones), but currently you return `BlobRef _ SerialisedBlob`. Also, is there anything else than a `BlobSpan` in the ref, e.g. to track whether the reference is still valid?

Reply:
The current types are:

* Public: https://github.com/IntersectMBO/lsm-tree/blob/d6cb82d02718347f0de1066be6abf62a97853fa2/src/Database/LSMTree/Common.hs#L264

* Internal: https://github.com/IntersectMBO/lsm-tree/blob/d6cb82d02718347f0de1066be6abf62a97853fa2/src/Database/LSMTree/Internal/BlobRef.hs#L41-L46

Whether we're going to need the `m` and `blob` parameters, specially for the internal type, that isn't clear. It should at least be in the public type so we know what type to deserialise the `SerialisedBlob` to, but maybe we can do without it internally

Reply:
I don't think the internal blob ref needs any type parameters except `fhandle` or probably `run` instead, to avoid cyclic dependencies and also in case we want to use it with something else than `Run`.

The public one could then be `data BlobRef m blob = forall h. Eq h => BlobRef (Internal.BlobRef (Internal.Run h))`.

I'm not sure how feasible it is to unify both representations, especially if we need existential type variables and constraints like `Eq h`.

Reply:
(has been addressed in #177)